sim/SIM-PROTOCOL.md

```diff
@@ -680,7 +680,12 @@ Mean Look or something), you will receive a message starting with:
 
 `|error|[Invalid choice]`
 
-This will tell you to send a different decision.
+This will tell you to send a different decision. If your previous choice
+revealed additional information, the error will contain an updated `REQUEST`
+object (in the same format as `|request|`) to base your decisions off of:
+
+`|error|[Invalid choice] MESSAGE|REQUEST`
```

Comment:
Hm.

You clearly need something in a choice error, to tell the random player AI "don't re-make the same choice, wait for a new request". But I don't like bundling the request here.

I think I'd prefer something like

```
|error|[Unavailable choice] MESSAGE
|request|NEWREQUEST
```

Reply:
Which I think would be backward-compatible across all existing sim uses?

Reply:
Having the two messages complicates `BattlePlayer#receiveLine`, no?

Reply:
Not any more than having a possible `REQUEST` term in `|error|`, in my view.

Reply:
Oh, I see what you mean. I'm surprised `receiveLine` isn't already designed to be multiline. But it would need to be sooner or later, anyway.

Reply:
Is that so? We expect clients to support lookahead?

Reply:
I'm not sure what you mean by "lookahead"? We expect clients to process messages line-by-line.

Reply:
`BattlePlayer` receive becomes:

```js
receive(chunk: string) {
	const lines = chunk.split('\n');
	if (lines.length >= 2 &&
		lines[0].startsWith('|error|[Invalid Choice]') &&
		lines[1].startsWith('|request|')) {
		const err = lines[0].slice(7);
		const request = lines[1].slice(9);
		this.receiveError(new Error(err), JSON.parse(request));
		const removed = lines.splice(0, 2);
		if (this.debug) console.log(removed);
	}

	for (const line of lines) {
		this.receiveLine(line);
	}
}
```

It needs to 'peek' (ie/ lookahead, a very common term in [parsing](https://en.wikipedia.org/wiki/Parsing#Lookahead)).

Reply:
Actually, I seem to recall a very angry comment from you in the client code when someone broke the 'line-by-line' requirement, so I don't think you want `BattlePlayer` to have to do this. If I'm mistaken and lookahead is not required, please show me the code `BattlePlayer`/`RandomPlayer` should be using with the separate `|error|` + `|request|` protocol.

Reply:
Oh. You're right, I don't want lookahead at all. That's why I suggested different names `[Unavailable choice]` vs `[Invalid choice]`, to make it clear whether or not a corrected request was forthcoming.

Reply:
This is also why I said I'd expect it to be backward compatible. Because `[Unavailable choice]` is already not considered a choice error by the client, so no additional code would be necessary.

Reply:
Something like `[Disabled choice]` might also work.

Reply:
I didn't notice the `[Unavailable choice]` distinction in your original comment or appreciate the significance.